### DISCUSSION TITLE
Why is it possible to pass a function as parameter to the add_system() function?

### DISCUSSION THREAD

**Discussion by Koettlitz** [bevyengine/bevy#4142] [Q&A] [answered]
The [`add_system()`](https://docs.rs/bevy/latest/bevy/app/struct.App.html#method.add_system) fn of the `App` struct expects a parameter of type `impl IntoSystemDescriptor<Params>`. How does a function match this type? I looked up the Implementors of [IntoSystemDescriptor](https://docs.rs/bevy/latest/bevy/ecs/schedule/trait.IntoSystemDescriptor.html) but I can't find any of the `Fn`-types. Can someone explain how this works?

**Comment by bjorn3** [comment, marked as answer]
There is a `impl<Params, S> IntoSystemDescriptor<Params> for S where S: IntoSystem<(), (), Params>` implementation. `IntoSystem` is implemented for all types implementing `SystemParamFunction` which is implemented for functions.

**Reply by Koettlitz** [reply]
Oh I just looked at the implementors of [`SystemParamFunction`](https://docs.rs/bevy/latest/bevy/ecs/system/trait.SystemParamFunction.html). It looks scary!

**Reply by bjorn3** [reply]
Yeah, the combination of handling the implementation using a macro and (ab)using the trait system of rustc (earlier iterations caused the compiler to crash) is indeed scary!

**Reply by Koettlitz** [reply]
But why does the following code not compile?
```rust
use bevy::prelude::{App, Commands, In};

fn main() {
    App::new().add_system(my_sys).run();
}

fn my_sys(_: In<usize>, _: Commands) {
    println!("Hugelbugel");
}
```

Why does it not match [this Blanket implementation](https://docs.rs/bevy/latest/bevy/ecs/system/trait.SystemParamFunction.html#impl-SystemParamFunction%3CInput%2C%20Out%2C%20(F0%2C)%2C%20InputMarker%3E) of [`SystemParamFunction`](https://docs.rs/bevy/latest/bevy/ecs/system/trait.SystemParamFunction.html)?

**Reply by bjorn3** [reply]
`In` is only allowed when you chain two systems together I believe. It will take the value of whatever the previous system in the chain returned.

**Comment by NickLarsenNZ** [comment]
I had the same question after realising I could put the params in any order (wondering how that was valid). Thanks for the answer @bjorn3
